readme.md
# Firewall 2.2

[![Latest Stable Version](https://img.shields.io/packagist/v/pragmarx/firewall.svg?style=flat-square)](https://packagist.org/packages/pragmarx/firewall) [![License](https://img.shields.io/badge/license-BSD_3_Clause-brightgreen.svg?style=flat-square)](LICENSE) [![Downloads](https://img.shields.io/packagist/dt/pragmarx/firewall.svg?style=flat-square)](https://packagist.org/packages/pragmarx/firewall) [![Code Quality](https://img.shields.io/scrutinizer/g/antonioribeiro/firewall.svg?style=flat-square)](https://scrutinizer-ci.com/g/antonioribeiro/firewall/?branch=master) [![Build](https://img.shields.io/scrutinizer/build/g/antonioribeiro/firewall.svg?style=flat-square)](https://scrutinizer-ci.com/g/antonioribeiro/firewall/?branch=master) [![Coverage](https://img.shields.io/scrutinizer/coverage/g/antonioribeiro/firewall.svg?style=flat-square)](https://scrutinizer-ci.com/g/antonioribeiro/firewall/?branch=master) [![StyleCI](https://styleci.io/repos/15290899/shield)](https://styleci.io/repos/15290899)

## Purpose

This a "soft-firewall" package. Its purpose is to help people prevent unauthorized access to routes **by IP address**. It is able to keep track of IPs, countries and hosts (dynamic ip), and redirect non-authorized users to, for instance, a "Coming Soon" page, while letting whitelisted IPs to have access to the entire site. It is now also able to detect and block attacks (too many requests) from single IPs or whole countries.

This package can prevent some headaches and help you block some access to your apps, but cannot replace firewalls and appliances, for attacks at the network level, you'll still need a real firewall. 

## Features

* Control access to routes and groups via black and white lists.
* Detect and block attacks to your application, from IP addresses or countries.
* Send Slack notifications in attack events.
* Allow whitelisted to access the whole site and send everyone else to a "coming soon page".
* Redirect blacklisted users to some other page.
* Use database or arrays to store IP lists.
* Whitelist your development machine using a dynamic DNS host name.
* Done using middleware, so you can protect/unprotect groups of routes.
* All features are available for hosts, IP addresses, ranges of IP addresses and whole countries.
* Super fast, less than 10ms increase in each request.
* Highly configurable.

## Concepts

### Blacklist

All IP addresses in those lists will no be able to access routes filtered by the blacklist filter.

### Whitelist

Those IP addresses, ranges or countries can

- Access blacklisted routes even if they are in a range of blacklisted IP addresses.
- Access 'allow whitelisted' filtered routes.
- If a route is filtered by the 'allow whitelisted' filter and the IP is not whitelisted, the request will be redirected to an alternative url or route name.

## Attack Detection

![attack](docs/attack.png)

Firewall is able to detect simple attacks to your page, by counting requests from the same IP or country. Just enable it on your `config/firewall.php` and, to receive notifications, configure the Slack service in `config/services.php`:
   
```php
'slack' => [
    'webhook_url' => env('SLACK_WEBHOOK_URL'),
],
```

and add the route notification method to your user model:

```php
/**
 * Route notifications for the Slack channel.
 *
 * @return string
 */
public function routeNotificationForSlack()
{
    return config('services.slack.webhook_url');
}
``` 

## IPs lists

IPs (white and black) lists can be stored in array, files and database. Initially database access to lists is disabled, so, to test your Firewall configuration you can publish the config file and edit the `blacklist` or `whitelist` arrays:

```php
'blacklist' => array(
    '127.0.0.1',
    '192.168.17.0/24'
    '127.0.0.1/255.255.255.255'
    '10.0.0.1-10.0.0.255'
    '172.17.*.*'
    'country:br'
    '/usr/bin/firewall/blacklisted.txt',
),
```

The file (for instance `/usr/bin/firewall/blacklisted.txt`) must contain one IP, range or file name per line, and, yes, it will search for files recursively, so you can have a file of files if you need:

```
127.0.0.2
10.0.0.0-10.0.0.100
/tmp/blacklist.txt
```

## Redirecting non-whitelisted IP addresses

Non-whitelisted IP addresses can be blocked or redirected. To configure redirection you'll have to publish the  `config.php` file and configure:

```php
'redirect_non_whitelisted_to' => 'coming/soon',
```

## Artisan Commands

You have access to the following commands:

#### Global

```
  firewall:cache:clear  Clear the firewall cache.
  firewall:list         List all IP address, white and blacklisted.
  firewall:updategeoip  Update the GeoIP database.
```

#### When database is enabled

```
  firewall:blacklist          Add an IP address to blacklist.
  firewall:clear              Remove all ip addresses from white and black lists.
  firewall:remove             Remove an IP address from white or black list.
  firewall:whitelist          Add an IP address to whitelist.
```

Those are results from `firewall:list`:

```
+--------------+-----------+-----------+
| IP Address   | Whitelist | Blacklist |
+--------------+-----------+-----------+
| 10.17.12.7   |           |     X     |
| 10.17.12.100 |     X     |           |
| 10.17.12.101 |     X     |           |
| 10.17.12.102 |     X     |           |
| 10.17.12.200 |           |     X     |
+--------------+-----------+-----------+
```

```
+-----------------------+-----------+-----------+
| IP Address            | Whitelist | Blacklist |
+-----------------------+-----------+-----------+
| 172.0.0.0-172.0.0.255 |           |     X     |
| country:br            |           |     X     |
| host:mypc.myname.com  |     X     |           |
+-----------------------+-----------+-----------+
```

## Facade

You can also use the `Firewall Facade` to manage the lists:

```php
$whitelisted = Firewall::isWhitelisted('10.17.12.1');
$blacklisted = Firewall::isBlacklisted('10.0.0.3');

Firewall::whitelist('192.168.1.1');
Firewall::blacklist('10.17.12.1', true); /// true = force in case IP is whitelisted
Firewall::blacklist('127.0.0.0-127.0.0.255');
Firewall::blacklist('200.212.331.0/28');
Firewall::blacklist('country:br');

if (Firewall::whichList($ip) !== false)  // returns false, 'whitelist' or 'blacklist'
{
    Firewall::remove($ip);
}
```

Return a blocking access response:

```php
return Firewall::blockAccess();
```

Suspicious events will be (if you wish) logged, so `tail` it:

```
php artisan tail
```

## Blocking Whole Countries

You can block a country by, instead of an ip address, pass `country:<2-letter ISO code>`. So, to block all Brazil's IP addresses, you do:

```
php artisan firewall:blacklist country:br
```

You will have to add this requirement to your `composer.json` file:

```
"geoip/geoip": "~1.14"
```

or

```
"geoip2/geoip2": "~2.0"
```

You need to enable country search on your firewall.php config file: 

```php
'enable_country_search' => true,
```

And you can schedule this command to update your cities GeoIp database regularly: 

```
php artisan firewall:updategeoip
```

You can find those codes here: [isocodes](http://www.spoonfork.org/isocodes.html)

## Session Blocking

You can block users from accessing some pages only for the current session, by using those methods:

```php
Firewall::whitelistOnSession($ip);
Firewall::blacklistOnSession($ip);
Firewall::removeFromSession($ip);
```

## Playground & Bootstrap App 

Click [here](http://pragmarx.com/firewall) to see it working and in case you need a help figuring out things, try [this repository](https://github.com/antonioribeiro/pragmarx.com). 

![playground](docs/playground.png)

## Installation

### Compatible with

- Laravel 4+ (version 1.*) 
- Laravel 5.0, 5.1, 5.2 and 5.3 (version 1.*)
- Laravel 5.4, 5.5, 5.6 and 5.7 (version 2.*)

### Installing

Require the Firewall package using [Composer](https://getcomposer.org/doc/01-basic-usage.md):

```
composer require pragmarx/firewall
```

* Laravel 5.5 and up

    You don't have to do anything else, this package uses Package Auto-Discovery's feature, and should be available as soon as you install it via Composer.

* Laravel 5.4 and below

    Add the Service Provider and the Facade to your app/config/app.php:

```php
PragmaRX\Firewall\Vendor\Laravel\ServiceProvider::class,
```

```php
'Firewall' => PragmaRX\Firewall\Vendor\Laravel\Facade::class,
```

Add middlewares to your app/Http/Kernel.php

```php
protected $routeMiddleware = [
    ...
    'fw-only-whitelisted' => \PragmaRX\Firewall\Middleware\FirewallWhitelist::class,
    'fw-block-blacklisted' => \PragmaRX\Firewall\Middleware\FirewallBlacklist::class,
    'fw-block-attacks' => \PragmaRX\Firewall\Middleware\BlockAttacks::class,
];
```

or 

```php
protected $middlewareGroups = [
    'web' => [
        ...
    ],

    'api' => [
        ...
    ],
    
    'firewall' => [
        \PragmaRX\Firewall\Middleware\FirewallBlacklist::class,
        \PragmaRX\Firewall\Middleware\BlockAttacks::class,
    ],
];
```

Then you can use them in your routes:

```php
Route::group(['middleware' => 'fw-block-blacklisted'], function () 
{
    Route::get('/', 'HomeController@index');
});
```

Or you could use both. In the following example the allow group will give free access to the 'coming soon' page and block or just redirect non-whitelisted IP addresses to another, while still blocking access to the blacklisted ones.

```php
Route::group(['middleware' => 'fw-block-blacklisted'], function () 
{
    Route::get('coming/soon', function()
    {
        return "We are about to launch, please come back in a few days.";
    });

    Route::group(['middleware' => 'fw-only-whitelisted'], function () 
    {
        Route::get('/', 'HomeController@index');
    });
});
```

**Note:** You can add other middleware you have already created to the new groups by simply 
adding it to the `fw-allow-wl` or `fw-block-bl` middleware group.

Migrate your database

```
php artisan migrate
```

**Warning:** If you already have a Firewall package installed and migrated, you need to update your migration name, in the `migrations` table, to `2014_02_01_311070_create_firewall_table`, otherwise the migrate command will fail tell you the table already exists.  

To publish the configuration file you'll have to:

**Laravel 4**

```
php artisan config:publish pragmarx/firewall
```

**Laravel 5**

```
php artisan vendor:publish --provider="PragmaRX\Firewall\Vendor\Laravel\ServiceProvider"
```

## TODO

- Tests, tests, tests.

## Author

[Antonio Carlos Ribeiro](http://twitter.com/iantonioribeiro) 

## License

Firewall is licensed under the BSD 3-Clause License - see the `LICENSE` file for details

## Contributing

Pull requests and issues are more than welcome.


changelog.md
# Change Log
All notable changes to this project will be documented in this file.

## [2.3.0] - 2019-09-11
### Added
- Laravel 6.0 support
- Support to identify IP behind Cloudflare
- Artisan command firewall:cache:clear
- Ability to set the log stack. New key in config: `'log_stack' => ['single']`

### Fixed
- Response in notification-only mode
- Migrations being execute when not in database mode
- Exception when trying to remove from database an ip inside a country

## [2.2.1] - 2018-07-31
### Fixed
- Fix whitelisted IP being blocked by AttackBlocker

## [2.0.2] - 2017-08-26
### Fixed
- Minor fixes

## [2.0.1] - 2017-08-21
### Fixed
- Minor fixes

## [2.0.0] - 2017-08-21
### Added
- Attack blocker

## [1.1.0] - 2017-08-20
### Added
- GeoIp2 database file updater artisan command

## [1.0.9] - 2017-08-13
### Added
- Laravel 5.5 autodiscovery
- Snippets to readme
### Changed
- Laravel 5.* install docs
### Fix
- Readme things
- Markdown headings

## [1.0.8] - 2017-01-31
### Updated
- To Laravel 5.4

## [1.0.7] - 2016-12-25
### Fixed
- Lexical problem

## [1.0.6] - 2016-12-24
### Updated
- Support package

## [1.0.5] - 2016-12-24
### Fixed
- Output on firewall:clear command
- Whitelisting problem

## [1.0.4] - 2016-12-24
### Added
- Ability to black/whitelist IPs for the session

## [1.0.3] - 2016-05-14
### Fixed
- Remove forgotten brace

## [1.0.2] - 2016-05-13
### Fixed
- Bug on PHP 7.0.2
- Deprecated Table helper (Symfony)

## [1.0.1] - 2016-04-18
### Changed
- Upgrade support package
### Fixed
- Middleware problem

## [1.0.0] - 2015-12-24
### Added
- Move to Middleware (BC)

## [1.0.0] - 2015-12-24
### Added
- Move to Middleware (BC)

## [0.5.4] - 2015-12-01
### Fixed
- Config publishing docs
- Minor fixes

## [0.5.3] - 2015-03
### Added
- Add support for GeoIP

## [0.5.2] - 2015-03
### Fixed
- Fix incorrectly using array as object
 
## [0.5.1] - 2014-02
### Fixed
- Minor fixes and improvements
 
## [0.5.0] - 2014-02-18
### Added
- Laravel 5 compatibility.
- Allow black and white lists to be stored in arrays, files or file of files, instead of a database table.
- Allow redirect_non_whitelisted_to to be a named route or url (to).
- Option to disable database lookup, disabled by default.
- Change log, according to http://keepachangelog.com/.
- Add cache for the list of IP addresses.

### Changed
- Removed unnecessary configuration options.
- Range search now defaults to enabled.

## [0.2.0] - 2014-07-08
### Fixed
- Migrator fixed
- Rename migrator


src/config/config.php
<?php

return [

    /*
     * Enable / disable firewall
     *
     */

    'enabled' => env('FIREWALL_ENABLED', true),

    /*
     * Whitelisted and blacklisted IP addresses, ranges, countries, files and/or files of files
     *
     *  Examples of IP address, hosts, country codes and CIDRs
     *      '127.0.0.1',
     *      '192.168.17.0/24'
     *      '127.0.0.1/255.255.255.255'
     *      '10.0.0.1-10.0.0.255'
     *      '172.17.*.*'
     *      'country:br'
     *      'host:google.com',
     *      storage_path().DIRECTORY_SEPARATOR.'blacklisted.txt', // a file with IPs, one per line
     */

    'blacklist' => [
    ],

    'whitelist' => [
    ],

    /*
     * Response action for blocked responses
     *
     */

    'responses' => [
        'blacklist' => [
            'code' => 403, // 200 = log && notify, but keep pages rendering

            'message' => null,

            'view' => null,

            'redirect_to' => null,

            'abort' => false, // return abort() instead of Response::make() - disabled by default
        ],

        'whitelist' => [
            'code' => 403, // 200 = log && notify, but keep pages rendering

            'message' => null,

            'view' => null,

            'redirect_to' => null,

            'abort' => false, // return abort() instead of Response::make() - disabled by default
        ],
    ],

    /*
     * Do you wish to redirect non whitelisted accesses to an error page?
     *
     * You can use a route name (coming.soon) or url (/coming/soon);
     *
     */

    'redirect_non_whitelisted_to' => null,

    /*
     * How long should we keep IP addresses in cache?
     *
     * This is a general client IP addresses cache. When the user hits your system his/her IP address
     * is searched and cached for the desired time. Finding an IP address contained in a CIDR
     * range (172.17.0.0/24, for instance) can be a "slow", caching it improves performance.
     *
     */

    'cache_expire_time' => 60, // minutes

    /*
     * How long should we keep lists of IP addresses in cache?
     *
     * This is the list cache. Database lists can take some time to load and process,
     * caching it, if you are not making frequent changes to your lists, may improve firewall speed a lot.
     */

    'ip_list_cache_expire_time' => 0, // minutes - disabled by default

    /*
     * Send suspicious events to log?
     *
     */

    'enable_log' => true,

    'log_stack' => null,

    /*
     * Search by range allow you to store ranges of addresses in
     * your black and whitelist:
     *
     *   192.168.17.0/24 or
     *   127.0.0.1/255.255.255.255 or
     *   10.0.0.1-10.0.0.255 or
     *   172.17.*.*
     *
     * Note that range searches may be slow and waste memory, this is why
     * it is disabled by default.
     *
     */

    'enable_range_search' => true,

    /*
     * Search by country range allow you to store country ids in your
     * your black and whitelist:
     *
     *   php artisan firewall:whitelist country:us
     *   php artisan firewall:blacklist country:cn
     *
     */

    'enable_country_search' => false,

    /*
     * Should Firewall use the database?
     */

    'use_database' => false,

    /*
     * Models
     *
     * When using the "eloquent" driver, we need to know which Eloquent models
     * should be used.
     *
     */

    'firewall_model' => 'PragmaRX\Firewall\Vendor\Laravel\Models\Firewall',

    /*
     * Session object binding in the IoC Container
     *
     * When blacklisting IPs for the current session, Firewall
     * will need to instantiate the session object.
     *
     */

    'session_binding' => 'session',

    /*
     * GeoIp2 database path.
     *
     * To get a fresh version of this file, use the command
     *
     *      php artisan firewall:updategeoip
     *
     */

    'geoip_database_path' => __DIR__.'/geoip', //storage_path('geoip'),

    /*
     * Block suspicious attacks
     */

    'attack_blocker' => [

        'enabled' => [
            'ip' => true,

            'country' => false,
        ],

        'cache_key_prefix' => 'firewall-attack-blocker',

        'allowed_frequency' => [

            'ip' => [
                'requests' => 50,

                'seconds' => 1 * 60, // 1 minute
            ],

            'country' => [
                'requests' => 3000,

                'seconds' => 2 * 60, // 2 minutes
            ],

        ],

        'action' => [

            'ip' => [
                'blacklist_unknown' => true,

                'blacklist_whitelisted' => false,
            ],

            'country' => [
                'blacklist_unknown' => false,

                'blacklist_whitelisted' => false,
            ],

        ],

        'response' => [
            'code' => 403, // 200 = log && notify, but keep pages rendering

            'message' => null,

            'view' => null,

            'redirect_to' => null,

            'abort' => false, // return abort() instead of Response::make() - disabled by default
        ],

    ],

    'notifications' => [
        'enabled' => true,

        'message' => [
            'title' => 'User agent',

            'message' => "A possible attack on '%s' has been detected from %s",

            'request_count' => [
                'title' => 'Request count',

                'message' => 'Received %s requests in the last %s seconds. Timestamp of first request: %s',
            ],

            'uri' => [
                'title' => 'First URI offended',
            ],

            'blacklisted' => [
                'title' => 'Was it blacklisted?',
            ],

            'user_agent' => [
                'title' => 'User agent',
            ],

            'geolocation' => [
                'title' => 'Geolocation',

                'field_latitude' => 'Latitude',

                'field_longitude' => 'Longitude',

                'field_country_code' => 'Country code',

                'field_country_name' => 'Country name',

                'field_city' => 'City',
            ],
        ],

        'route' => '',

        'from' => [
            'name' => 'Laravel Firewall',

            'address' => 'firewall@mydomain.com',

            'icon_emoji' => ':fire:',
        ],

        'users' => [
            'model' => PragmaRX\Firewall\Vendor\Laravel\Models\User::class,

            'emails' => [
                'admin@mydomain.com',
            ],
        ],

        'channels' => [
            'slack' => [
                'enabled' => true,
                'sender'  => PragmaRX\Firewall\Notifications\Channels\Slack::class,
            ],

            'mail' => [
                'enabled' => true,
                'sender'  => PragmaRX\Firewall\Notifications\Channels\Mail::class,
            ],
        ],
    ],
];
